### PR TITLE
SMS Survey Refactor: 2

### PR DESCRIPTION
**Pull request by gcapalbo**
The next step in the SMS survey refactor, this locks out all places in the code that edit sms survey sessions and introduces a new process that polls for sessions that have actions due, spawning celery tasks to handle the action.

The last step in this refactor will be to remove all sms survey functionality from the reminders framework (except for initiating the survey), and letting the new celery task handle all survey-related actions after that.

review with `?w=1`

@proteusvacuum

### CODE REVIEW

File: corehq/apps/smsforms/management/commands/handle_survey_actions.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        for domain, connection_id, session_id, current_action_due in self.get_survey_sessions_due_for_action():
+            if skip_domain(domain):
+                continue
+
+            enqueue_lock = self.get_enqueue_lock(session_id, current_action_due)
+            if enqueue_lock.acquire(blocking=False):
+                handle_due_survey_action.delay(domain, connection_id, session_id)
+
+    def handle(self, **options):
+        while True:
```
Comment (proteusvacuum) [inline corehq/apps/smsforms/management/commands/handle_survey_actions.py]: why is this in a `while True`? I'm not seeing anything that will ever break out of this loop.

File: corehq/apps/smsforms/tests/test_sql_session.py [hunk 2]
```diff
@@ -141,6 +149,99 @@ def test_get_open_sms_session_one_result(self):
         session = SQLXFormsSession.get_open_sms_session(domain, contact)
         self.assertEqual(new_session.session_id, session.session_id)
 
+    @patch('corehq.apps.smsforms.models.utcnow')
```
Comment (proteusvacuum) [inline corehq/apps/smsforms/tests/test_sql_session.py]: Have we ever looked at using something like https://github.com/spulec/freezegun/ ? Making a special utcnow function seems pretty clunky.

File: corehq/apps/smsforms/tests/test_sql_session.py [hunk 3]
```diff
@@ -7,9 +7,17 @@
 from corehq.apps.sms.handlers.form_session import get_single_open_session_or_close_multiple
 from corehq.apps.smsforms.models import SQLXFormsSession, XFORMS_SESSION_TYPES, XFORMS_SESSION_SMS, \
     XFORMS_SESSION_IVR
+from mock import patch
 from six.moves import range
 
 
+class MockObject(object):
```
Comment (proteusvacuum) [inline corehq/apps/smsforms/tests/test_sql_session.py]: I think `mock.Mock` does pretty much exactly this. 

```
In [1]: import mock

In [2]: mock.Mock
Out[2]: mock.mock.Mock

In [3]: x = mock.Mock(get_id='contact_id')

In [4]: x.get_id
Out[4]: 'contact_id'
```
Or is there some other reason you made your own?

**Review by proteusvacuum — APPROVED** [review]

File: corehq/apps/smsforms/management/commands/handle_survey_actions.py (hunk 1 above)
Comment (gcapalbo) [reply corehq/apps/smsforms/management/commands/handle_survey_actions.py]: this is run in a supervisor process: https://github.com/dimagi/commcare-cloud/pull/1280

the loop breaks when supervisor sends a sigterm

File: corehq/apps/smsforms/tests/test_sql_session.py (hunk 2 above)
Comment (gcapalbo) [reply corehq/apps/smsforms/tests/test_sql_session.py]: i haven't looked into using a third party tool to do this, but in general mocking datetimes is not a straightforward thing to do

File: corehq/apps/smsforms/tests/test_sql_session.py (hunk 3 above)
Comment (gcapalbo) [reply corehq/apps/smsforms/tests/test_sql_session.py]: updated: https://github.com/dimagi/commcare-hq/pull/19161/commits/ca0c118a20d5ec1df3ca531800705d9e99122549

File: corehq/apps/smsforms/management/commands/handle_survey_actions.py (hunk 1 above)
Comment (proteusvacuum) [reply corehq/apps/smsforms/management/commands/handle_survey_actions.py]: Oh I see. I'd missed that. Thanks for the explanation.

**Pull request merged**